src/app/shared/file-preview/model-file-preview/model-file-preview.service.ts

```diff
@@ -0,0 +1,212 @@
+            }
+
+            // vertices and normals
+            for (const vertex of [faceA, faceB, faceC]) {
+                vertices.push(model.verticesX[vertex]);
+                vertices.push(-model.verticesY[vertex]);
+                vertices.push(model.verticesZ[vertex]);
+
+                const vertexNormal = model.vertexNormals[vertex];
+                normals.push(vertexNormal.x);
+                normals.push(vertexNormal.y);
+                normals.push(vertexNormal.z);
+            }
+
+            // face colors
+            switch (faceType) {
+                case 0: // shaded face
+                case 2: // textured shaded face
+                    // TODO Apply texture color
+                    // TODO Apply shadow
+                    const hsbColor = model.faceColors[i];
+                    const shadowedColorX = hsbColor; // ColorUtils.method709(hsbColor, model.faceColorsX[i]);
+                    const shadowedColorY = hsbColor; // ColorUtils.method709(hsbColor, model.faceColorsY[i]);
+                    const shadowedColorZ = hsbColor; // ColorUtils.method709(hsbColor, model.faceColorsZ[i]);
+                    const colorX = new THREE.Color(ColorUtils.hsbToRgb(shadowedColorX));
+                    const colorY = new THREE.Color(ColorUtils.hsbToRgb(shadowedColorY));
+                    const colorZ = new THREE.Color(ColorUtils.hsbToRgb(shadowedColorZ));
+
+                    for(const rgb of [colorX, colorY, colorZ]) {
+                        colors.push(rgb.r);
+                        colors.push(rgb.g);
+                        colors.push(rgb.b);
+                    }
+                    break;
+                case 1: // flat face
+                case 3: // textured flat face
+                    // TODO Apply texture color
+                    const color = new THREE.Color(ColorUtils.hsbToRgb(model.faceColorsX[i]));
+                    for(const rgb of [color, color, color]) {
+                        colors.push(rgb.r);
+                        colors.push(rgb.g);
+                        colors.push(rgb.b);
+                    }
+                    break;
+            }
+
+            // TODO Multi materials (textures)
+        }
+        geometry.setAttribute('position', new THREE.BufferAttribute(new Float32Array(vertices), 3));
+        geometry.setAttribute('normal', new THREE.BufferAttribute(new Float32Array(normals), 3));
+        geometry.setAttribute('color', new THREE.BufferAttribute(new Float32Array(colors), 3));
+
+        const mesh = new THREE.Mesh(geometry, materials[0]);
+        mesh.rotateY(Math.PI);
+        const scale = ModelFilePreviewService.MODEL_SCALE;
+        mesh.scale.set(scale, scale, scale);
+
+        this.rsModelMesh = mesh;
+        this.scene.add(mesh);
+    }
+
+    public animate(): void {
+        // We have to run this outside angular zones,
+        // because it could trigger heavy changeDetection cycles.
+        this.ngZone.runOutsideAngular(() => {
```

Comment:
💯 beautiful

Reply:
I copied that from: https://github.com/JohnnyDevNull/ng-three-template xD.